src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java

```diff
@@ -0,0 +1,268 @@
+import com.ctre.phoenix.motorcontrol.FeedbackDevice;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.SupplyCurrentLimitConfiguration;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.ctre.phoenix.sensors.SensorInitializationStrategy;
+import edu.wpi.first.math.Nat;
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.controller.LinearQuadraticRegulator;
+import edu.wpi.first.math.estimator.KalmanFilter;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.math.numbers.N1;
+import edu.wpi.first.math.system.LinearSystem;
+import edu.wpi.first.math.system.LinearSystemLoop;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.math.system.plant.LinearSystemId;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.subsystems.UnitModel;
+import frc.robot.utils.utils.SwerveModuleConfigBase;
+import frc.robot.utils.utils.Units;
+import webapp.FireLog;
+
+/**
+ * This subsystem represents a single Swerve module and is responsible for the basic operation of a module, such as,
+ * rotating to a specific angle, driving at a specific rate, and other convenience features for tuning the coefficients.
+ */
+public class SwerveModule extends SubsystemBase {
+    private final WPI_TalonFX driveMotor;
+    private final WPI_TalonSRX angleMotor;
+    private final UnitModel driveUnitModel;
+    private final UnitModel angleUnitModel;
+
+    private final SwerveModuleConfigBase config;
+    private LinearSystemLoop<N1, N1, N1> stateSpace;
+    private double currentTime, lastTime;
+    private double lastJ;
+
+    public SwerveModule(SwerveModuleConfigBase config) {
+        this.config = config;
+        driveMotor = new WPI_TalonFX(config.driveMotorPort());
+        angleMotor = new WPI_TalonSRX(config.angleMotorPort());
+        driveUnitModel = new UnitModel(Constants.SwerveDrive.DRIVE_MOTOR_TICKS_PER_METER);
+        angleUnitModel = new UnitModel(Constants.SwerveDrive.ANGLE_MOTOR_TICKS_PER_RADIAN);
+        stateSpace = constructVelocityLinearSystem(config.j());
+        stateSpace.reset(VecBuilder.fill(Units.metersPerSecondToRps(getVelocity(), Constants.SwerveDrive.WHEEL_RADIUS)));
+        lastJ = config.j();
+
+        // configure feedback sensors
+        angleMotor.configSelectedFeedbackSensor(FeedbackDevice.Analog, 0, Constants.TALON_TIMEOUT);
+        angleMotor.configFeedbackNotContinuous(false, Constants.TALON_TIMEOUT);
+
+        angleMotor.setNeutralMode(NeutralMode.Brake);
+        driveMotor.setNeutralMode(NeutralMode.Brake);
+
+        // inversions
+        driveMotor.setInverted(config.driveMotorInverted());
+        angleMotor.setInverted(config.angleMotorInverted());
+
+        driveMotor.setSensorPhase(config.driveMotorSensorPhase());
```

Comment:
it turns out that we don't need to config the sensor phase of TalonFxs ([referece](https://phoenix-documentation.readthedocs.io/en/latest/ch14_MCSensor.html#sensor-phase))